kafka-connect-s3/src/test/java/io/confluent/connect/s3/DataWriterByteArrayTest.java

```diff
@@ -0,0 +1,228 @@
+    partitioner.configure(parsedConfig);
+    format = new ByteArrayFormat(storage);
+    s3.createBucket(S3_TEST_BUCKET_NAME);
+    assertTrue(s3.doesBucketExist(S3_TEST_BUCKET_NAME));
+  }
+
+  @After
+  @Override
+  public void tearDown() throws Exception {
+    super.tearDown();
+    localProps.clear();
+  }
+
+  @Test
+  public void testNoSchema() throws Exception {
+    localProps.put(S3SinkConnectorConfig.FORMAT_CLASS_CONFIG, ByteArrayFormat.class.getName());
+    setUp();
+    task = new S3SinkTask(connectorConfig, context, storage, partitioner, format, SYSTEM_TIME);
+
+    List<SinkRecord> sinkRecords = createByteArrayRecordsWithoutSchema(7 * context.assignment().size(), 0, context.assignment());
+    task.put(sinkRecords);
+    task.close(context.assignment());
+    task.stop();
+
+    long[] validOffsets = {0, 3, 6};
+    verify(sinkRecords, validOffsets, context.assignment(), ".bin");
+  }
+
+  @Test
+  public void testCustomExtensionAndLineSeparator() throws Exception {
+    String extension = ".SEPARATORjson";
+    localProps.put(S3SinkConnectorConfig.FORMAT_CLASS_CONFIG, ByteArrayFormat.class.getName());
+    localProps.put(S3SinkConnectorConfig.FORMAT_BYTEARRAY_LINE_SEPARATOR_CONFIG, "SEPARATOR");
+    localProps.put(S3SinkConnectorConfig.FORMAT_BYTEARRAY_EXTENSION_CONFIG, extension);
+    setUp();
+    task = new S3SinkTask(connectorConfig, context, storage, partitioner, format, SYSTEM_TIME);
+
+    List<SinkRecord> sinkRecords = createByteArrayRecordsWithoutSchema(7 * context.assignment().size(), 0, context.assignment());
+    task.put(sinkRecords);
+    task.close(context.assignment());
+    task.stop();
+
+    long[] validOffsets = {0, 3, 6};
+    verify(sinkRecords, validOffsets, context.assignment(), extension);
+  }
+
+  protected List<SinkRecord> createByteArrayRecordsWithoutSchema(int size, long startOffset, Set<TopicPartition> partitions) {
+    String key = "key";
+    int ibase = 12;
+
+    List<SinkRecord> sinkRecords = new ArrayList<>();
+    for (long offset = startOffset, total = 0; total < size; ++offset) {
+      for (TopicPartition tp : partitions) {
+        byte[] record = ("{\"schema\":{\"type\":\"struct\",\"fields\":[ " +
+                            "{\"type\":\"boolean\",\"optional\":true,\"field\":\"booleanField\"}," +
+                            "{\"type\":\"int32\",\"optional\":true,\"field\":\"intField\"}," +
+                            "{\"type\":\"int64\",\"optional\":true,\"field\":\"longField\"}," +
+                            "{\"type\":\"string\",\"optional\":false,\"field\":\"stringField\"}]," +
+                            "\"payload\":" +
+                            "{\"booleanField\":\"true\"," +
+                            "\"intField\":" + String.valueOf(ibase) + "," +
+                            "\"longField\":" + String.valueOf((long) ibase) + "," +
+                            "\"stringField\":str" + String.valueOf(ibase) +
+                            "}}").getBytes();
```

Comment:
I copied this test verbatim from the JSON tests, so this message doesn't make much sense here.

I tried out a change where instead of this message, we just fill a byte array with random bytes (and make sure not to include the separator bytes). I think that exposed a logic error in the existing JSON tests. This is structured with TopicPartition as the inner loop, but `verify` puts TopicPartition as the outer loop, so expectedRecords and the reconstituted records don't match up.

It looks like the JSON tests only succeed because all the messages have the same value.

Do you have any opinions about what payloads we should be generating here to fully exercise the new format? Is the test as written right now fine for the scope of this PR? Or would you prefer that we go ahead and modify the JSON test logic inside this PR and factor out this method to be shared between the JSON and byte array tests?

Reply:
From what I can remember having the same message was an intentional tradeoff. We should change this in both cases, but that happen in a follow up PR.